create script to upgrade metalk8s cluster

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'salt', 'upgrade'

**Context**: 

Issue: #1403 

**Summary**:

- Add a buildchain task to copy the upgrade script
- Add an upgrade script
- Fix documentation according to new upgrade script

**Acceptance criteria**: 

- Be able to use the upgrade script to upgrade MetalK8s
---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1403

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1402-create-script-to-upgrade-metalk8s-cluster` into
`development/2.0`
* [w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.1` 
* [w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

slaperche-scality commented on buildchain/buildchain/iso.py:
```diff
@@ -126,6 +126,21 @@ def task__iso_render_bootstrap() -> types.TaskDict:
+        'targets': [dest],
+        'task_dep': [__iso_mkdir_root],
+        'file_dep': [src],
+        'uptodate': [True],
```
Nope.
Why do you want it to be always true?

If the script is modified, you want to copy it again.

bert-e commented:
# History mismatch



Merge commit #73fcb009e943731a60eca936ac5ee87d23879086 on the integration branch
`w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster` is merging a branch which is neither the current
branch `improvement/1402-create-script-to-upgrade-metalk8s-cluster` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1402-create-script-to-upgrade-metalk8s-cluster` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented on buildchain/buildchain/iso.py:
```diff
@@ -126,6 +126,21 @@ def task__iso_render_bootstrap() -> types.TaskDict:
     ).task
 
 
+def task__add_upgrade() -> type.TaskDict:
```
We now have a lot of duplication between `task__iso_add_iso_manager`, `task__add_upgrade` and `task__iso_add_node_manifest`.
We should refactor this (but you don't need to do it here, I'll open a debt issue for that)

Ebaneck replied:
ok

slaperche-scality replied:
Tracked in #1439

slaperche-scality commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,243 @@
+#
+set -e
+set -u
+# The trickery below is to only run `set -o pipefail`, which is a Bash'ism,
```
I don't think you need this trick, as you are explicitly using `#!/bin/bash`

slaperche-scality commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,243 @@
+# support `-o pipefail` (though must support `-e` and `-u` per POSIX).
+if test -z "$(type -p)"; then set -o pipefail; fi
+
+# set -o nounset # Treat unset variables as an error
```
Why is this commented?

Either un-comment or remove the code.

slaperche-scality commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,243 @@
+trap cleanup EXIT
+
+
+run_quiet() {
```
It seems to me that we have a lot of copy/paste between our scripts.

We may want (not in this PR, it's out of scope) to have a shell script library with our common function inside, no?

sayf-eddine-scality replied:
yes I agree

slaperche-scality replied:
Tracked in #1441

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,243 @@
+#ANSI color codes:
+RED='\033[1;31m'
+GREEN='\033[1;32m'
+NOCOLOR='\033[0m'
```
As we use no color in bootstrap script. I suggest to not use it here

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,243 @@
+  Usage :  ${__ScriptName} <options>
+  Options:
+    -h | --help              Display HELP options
+    -v | --version           Show current installed version
```
To be consistent `-v` is for verbose

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,237 @@
+
+__read_version()
+{
+  "${SALT_CALL}" slsutil.renderer string="{{ pillar.metalk8s.nodes[grains.id].version }}"
```
Indentation here (different from others functions)

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,235 @@
+
+
+orchestrate_upgrade() {
+
```
Remove this extra line

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,235 @@
+}
+
+__orchestrate_dryrun(){
+
```
Same

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,235 @@
+    run "Performing Pre-Upgrade prechecks" \
+    "${SALT_MASTER_CALL[@]}" salt-run state.orchestrate metalk8s.orchestrate.upgrade.precheck saltenv=metalk8s-$TO_VERSION \
+        pillar="{'orchestrate': {'dest_version': '$TO_VERSION'}}"
+
```
Ditto

Ebaneck commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/1402-create-script-to-upgrade-metalk8s-cluster` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`improvement/1402-create-script-to-upgrade-metalk8s-cluster`).

```bash
 $ git fetch
 $ git checkout origin/improvement/1402-create-script-to-upgrade-metalk8s-cluster
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/1402-create-script-to-upgrade-metalk8s-cluster
```

slaperche-scality commented on buildchain/buildchain/iso.py:
```diff
@@ -79,6 +79,7 @@ def task_populate_iso() -> types.TaskDict:
             '_iso_render_bootstrap',
```
I think the change to this files belong to the other commit (`buildchain: Add task to copy upgrade script`)

Ebaneck replied:
I found the culprit...It is fixed now

slaperche-scality replied:
Hum, I'm afraid that something went wrong.

The changes are still in this commit, and you did a merge in your PR (which we usually avoid inside a PR).

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1402-create-script-to-upgrade-metalk8s-cluster` into
`development/2.0`
* [w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.1` 
* [w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

* @teddyandrieux

bert-e commented:
# History mismatch



Merge commit #f1d0ed1eb586d7cf8d234c0bb85e425ad76d9ff0 on the integration branch
`w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster` is merging a branch which is neither the current
branch `improvement/1402-create-script-to-upgrade-metalk8s-cluster` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1402-create-script-to-upgrade-metalk8s-cluster` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck committed:
salt: Add bash script to support upgrade

This commit adds a bash script for upgrading MetalK8s stack.

Initially, the procedure was not user friendly. Now, upgrading
requires just a single command.

Ebaneck committed:
doc: Improve documentation with upgrade script

This commit adds improvement to the documentation
regarding the new upgrade script.

Launching upgrade requires just this script

Ebaneck committed:
buildchain: Add copy-task for upgrade script

This commit adds a task to copy the upgrade script
while packaging the MetalK8s iso file.

slaperche-scality commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,203 @@
+#          that make up the stack.
+#
+set -e
+set -u
```
I think we want to keep the `set -o pipefail` (you removed what was around and that's good, you just removed a bit too much :slightly_smiling_face: )

slaperche-scality reviewed (dismissed):
LGTM (except one comment), will let the Bash expert review the script itself.

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,203 @@
+    return 1
+}
+
+__read_version() {
```
I think this function (and the argument `--version`) is not relevant for the moment except if you show the version of all nodes because here you will only have the version of the bootstrap node

Ebaneck replied:
ok

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,203 @@
+
+orchestrate_upgrade() {
+
+    run "Performing preliminary bootstrap node upgrade in local mode(this may take a while)" \
```
```suggestion
    run "Performing bootstrap stuff upgrade(this may take a while)" \
```
This one it's only about upgrading bootstrap stuff that will not upgrade apiserver, etc, ... of the bootstrap node.
Also I think we don't need to tell the user that it's a local mode execution

Ebaneck replied:
Yeah I can make this change..

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,203 @@
+
+    run "Performing Pre-Upgrade prechecks" \
+    "${SALT_MASTER_CALL[@]}" salt-run state.orchestrate metalk8s.orchestrate.upgrade.precheck saltenv=metalk8s-"$TO_VERSION" \
+        pillar="{'orchestrate': {'dest_version': '$TO_VERSION'}}"
```
You don't need to run this state as it's in a script this state is already included in the second one and it's only some checking

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,203 @@
+        pillar="{'orchestrate': {'dest_version': '$TO_VERSION'}}"
+
+    run "Verifying that upgrade was successful(New version installed)" \
+    "${SALT_CALL}" slsutil.renderer string="{{ pillar.metalk8s.nodes[grains.id].version }}"
```
Btw this is not "Verifying" at this will only print the version of the bootstrap node and not compare it with the destination one.

And btw you can't compare it with the destination one as during upgrade (and downgrade) some nodes may not change version.

For example if you do an upgrade to 2.2.0 and you have one node in 2.2.1 this node will stay in 2.2.1 even if destination is 2.2.0 because it's upgrade

Ebaneck commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1402-create-script-to-upgrade-metalk8s-cluster` into
`development/2.0`
* [w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.1` 
* [w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.2` 
* [w/2.3/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.3` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

* @teddyandrieux

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

sayf-eddine-scality commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,187 @@
+        pillar="{'orchestrate': {'dest_version': '$TO_VERSION'}}"
+}
+
+__orchestrate_dryrun() {
```
dry run is just adding `test=True` to the salt calls not a separate function by itself, you need to flip this key when `-d` is provided as argument

TeddyAndrieux replied:
No. For Upgrade/downgrade I think it's better to run the precheck state and that it
The `test=true` will not give you lots of info as it's depend on the version your set on each node and btw as it's test that will set nothing.

bert-e commented:
# History mismatch



Merge commit #4e8957b6eb374e90b2c14e9a1a762b79755d91ef on the integration branch
`w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster` is merging a branch which is neither the current
branch `improvement/1402-create-script-to-upgrade-metalk8s-cluster` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1402-create-script-to-upgrade-metalk8s-cluster` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,187 @@
+    case "$1" in
+        -d|--dry-run)
+            VERBOSE=1;
+            TO_VERSION="$2"
```
Hmm, no don't really like that.
All args should be parsed before you run something.

Get `__orchestrate_dryrun` and `orchestrate_upgrade` out of the `while` loop.

bert-e commented:
# History mismatch



Merge commit #39461ba28cc7c48b2081e1593ffda3d72167a742 on the integration branch
`w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster` is merging a branch which is neither the current
branch `improvement/1402-create-script-to-upgrade-metalk8s-cluster` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1402-create-script-to-upgrade-metalk8s-cluster` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,202 @@
+    case "$1" in
+        -p|--pre-check)
+            VERBOSE=1;
+            TO_VERSION="$2"
```
No, do not store version here. Only set a variable `DRY_RUN` for example to `1`

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,202 @@
+            shift;;
+        -t|--to)
+            TO_VERSION="$2"
+            UPGRADE=1
```
This one should only store the version that it. And after you will execute something in function if you have `DRY_RUN` set to `1` or not

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,202 @@
+fi
+}
+
+main
```
Honnestly I don't think you really need to do a `main` function for a bash script like that

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,202 @@
+        -v|--verbose)
+            VERBOSE=1;
+            shift;;
+        -t|--to)
```
Also I personnally not really like the argument `--to` consider use something like `--destination-version` or something else but consistent with @sayf-eddine-scality PR

Ebaneck replied:
ok

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,184 @@
+}
+
+if [ -n "$DESTINATION_VERSION" ]; then
+        __orchestrate_precheck
```
It's useless to run precheck if you are not in dry run mode.
If dry_run mode then only precheck else only upgrade

TeddyAndrieux replied:
Oh or ok lets keep it so that you have a small output if you can't Upgrade

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,184 @@
+__orchestrate_precheck() {
+    SALT_MASTER_CALL=(crictl exec -i "$(get_salt_container)")
+
+    run "Performing Pre-Upgrade prechecks" \
```
"Pre-Upgrade checks" or "Upgrade prechecks" as you want :)

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,184 @@
+        pillar="{'orchestrate': {'dest_version': '$DESTINATION_VERSION'}}"
+}
+
+if [ -n "$DESTINATION_VERSION" ]; then
```
Like in #1446 you should do this check in the "precheck" function

Ebaneck commented:
@bert-e  reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1402-create-script-to-upgrade-metalk8s-cluster` into
`development/2.0`
* [w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.1` 
* [w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.2` 
* [w/2.3/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.3` 
* [w/2.4/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,191 @@
+if [ $DRY_RUN -eq 1 ]; then
+    __orchestrate_precheck
+    else
+        __orchestrate_upgrade
```
No I answered this one lets keep the precheck always . Try to do something like in @sayf-eddine-scality PR
```
run "Performing Pre-Downgrade checks" precheck_downgrade
[ $DRY_RUN -eq 1 ] && exit 0
run "Launching the downgrade" launch_downgrade
run "Downgrading bootstrap" downgrade_bootstrap
```
Replacing Downgrade by Upgrade of course.

Also like in @sayf-eddine-scality PR try to use `run` the command here.
Please sync together to have the same "template of script"
I mean both script should be very similar.

Ebaneck replied:
Well noted

bert-e commented:
# History mismatch



Merge commit #5169ceec6bf7b91491f1e9870ff6dbdbd41e3390 on the integration branch
`w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster` is merging a branch which is neither the current
branch `improvement/1402-create-script-to-upgrade-metalk8s-cluster` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1402-create-script-to-upgrade-metalk8s-cluster` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1402-create-script-to-upgrade-metalk8s-cluster` into
`development/2.0`
* [w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.1` 
* [w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.2` 
* [w/2.3/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.3` 
* [w/2.4/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

Ebaneck commented:
@bert-e wait

TeddyAndrieux commented on buildchain/buildchain/iso.py:
```diff
@@ -86,6 +86,7 @@ def task_populate_iso() -> types.TaskDict:
         ],
     }
 
+
```
Nop, remove this new line

TeddyAndrieux commented on docs/operation/upgrade.rst:
```diff
@@ -35,57 +25,9 @@ Upgrade Steps
+   .. code::
 
-        salt-call slsutil.renderer string="{{ pillar.metalk8s.nodes[grains.id].version }}"
+     /srv/scality/metalk8s-X.X.X/upgrade.sh --to <destination_version>
```
It's not longer `--to`

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,190 @@
+    -v | --verbose               Run in verbose mode
+EOF
+exit
+}
```
To be consistent with downgrade script
```
_usage() {
    echo "downgrade.sh [options]"
    echo "Options:"
    echo "--destination-version "
    echo "   <destination-version>:        Destination version to upgrade to"
    echo "-l/--log-file <logfile_path>:    Path to log file"
    echo "-v/--verbose:                    Run in verbose mode"
    echo "-d/--dry-run:                    Run actions in dry run mode"
    echo "-h/--help:                       Show this help menu"
}

```

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,190 @@
+            shift 2;;
+        -h|--help)
+            __usage
+            shift;;
```
Add `exit` here

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,190 @@
+            echo "K.O   Incorrect arguments" 1>&2
+            exit 1 ;;
+    esac
+done
```
```
while (( "$#" )); do
  case "$1" in
    -d|--dry-run)
      DRY_RUN=1
      shift
      ;;
    -v|--verbose)
      VERBOSE=1
      shift
      ;;
    --destination-version)
      DESTINATION_VERSION="$2"
      shift 2
      ;;
    -l|--log-file)
      LOGFILE="$2"
      shift 2
      ;;
    -h|--help)
      _usage
      exit
      ;;
    *) # unsupported flags
      echo "Error: Unsupported flag $1" >&2
      _usage
      exit 1
      ;;
  esac
done
```
Please really stick to the same script as downgrade one so that it will be easier to maintain

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,180 @@
+    "${SALT_CALL}" --local state.sls metalk8s.roles.bootstrap saltenv=metalk8s-"$DESTINATION_VERSION" \
+        pillar="{'metalk8s': {'endpoints': $(salt-call --out txt pillar.get metalk8s:endpoints \
+        | cut -c 8-)}}"
+
```
Remove this empty line

Ebaneck replied:
Hmmm... do you have a linting tool that finds this automatically?? Sure vscode should have a plugin for this.

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,180 @@
+}
+
+upgrade_bootstrap () {
+    "${SALT_CALL}" --local state.sls metalk8s.roles.bootstrap saltenv=metalk8s-"$DESTINATION_VERSION" \
```
You should quote the whole saltenv not only the version
```suggestion
    "${SALT_CALL}" --local state.sls metalk8s.roles.bootstrap saltenv="metalk8s-$DESTINATION_VERSION" \
```

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,180 @@
+launch_upgrade () {
+    SALT_MASTER_CALL=(crictl exec -i "$(get_salt_container)")
+
+    "${SALT_MASTER_CALL[@]}" salt-run state.orchestrate metalk8s.orchestrate.upgrade saltenv=metalk8s-"$DESTINATION_VERSION" \
```
```suggestion
    "${SALT_MASTER_CALL[@]}" salt-run state.orchestrate metalk8s.orchestrate.upgrade saltenv="metalk8s-$DESTINATION_VERSION" \
```

TeddyAndrieux commented on scripts/upgrade.sh:
```diff
@@ -0,0 +1,180 @@
+    fi
+    SALT_MASTER_CALL=(crictl exec -i "$(get_salt_container)")
+
+    "${SALT_MASTER_CALL[@]}" salt-run state.orchestrate metalk8s.orchestrate.upgrade.precheck saltenv=metalk8s-"$DESTINATION_VERSION" \
```
```suggestion
    "${SALT_MASTER_CALL[@]}" salt-run state.orchestrate metalk8s.orchestrate.upgrade.precheck saltenv="metalk8s-$DESTINATION_VERSION" \
```

Ebaneck commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **wait**

TeddyAndrieux reviewed (approved):
LGTM

Ebaneck commented:
@bert-e create_pull_requests

Ebaneck commented:
@bert-e approve

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1402-create-script-to-upgrade-metalk8s-cluster` into
`development/2.0`
* [w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.1` 
* [w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.2` 
* [w/2.3/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.3` 
* [w/2.4/improvement/1402-create-script-to-upgrade-metalk8s-cluster](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/improvement/1402-create-script-to-upgrade-metalk8s-cluster) will be merged into `development/2.4` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye ebaneck.
